banyand/liaison/grpc/property.go

```diff
@@ -357,34 +357,76 @@ func (ps *propertyServer) Query(ctx context.Context, req *propertyv1.QueryReques
 		req.Limit = 100
 	}
 
-	nodeProperties, trace, err := ps.queryProperties(ctx, req)
+	nodeProperties, groups, trace, err := ps.queryProperties(ctx, req)
 	if err != nil {
 		return nil, err
 	}
-	res := make(map[string]*propertyWithMetadata)
-	for _, nodeWithProperties := range nodeProperties {
+	res := make(map[string]*propertyWithCount)
+	shouldDeleteOlderProperties := make([][]byte, 0)
+	for n, nodeWithProperties := range nodeProperties {
 		for _, propertyMetadata := range nodeWithProperties {
 			entity := propertypkg.GetEntity(propertyMetadata.Property)
 			cur, ok := res[entity]
 			if !ok {
-				res[entity] = propertyMetadata
+				res[entity] = newPropertyWithCounts(propertyMetadata, n)
 				continue
 			}
-			if cur.Metadata.ModRevision < propertyMetadata.Metadata.ModRevision {
-				res[entity] = propertyMetadata
-				// TODO(mrproliu) handle the case where the property detected multiple versions
+			switch {
+			case cur.Metadata.ModRevision < propertyMetadata.Metadata.ModRevision: // newer revision
+				res[entity] = newPropertyWithCounts(propertyMetadata, n)
+				shouldDeleteOlderProperties = append(shouldDeleteOlderProperties, propertypkg.GetPropertyID(cur.Property))
+			case cur.Metadata.ModRevision == propertyMetadata.Metadata.ModRevision: // same revision
+				cur.AddExistNode(n)
+			default: // cur.Metadata.ModRevision > propertyMetadata.Metadata.ModRevision
+				shouldDeleteOlderProperties = append(shouldDeleteOlderProperties, propertypkg.GetPropertyID(propertyMetadata.Property))
 			}
 		}
 	}
+	if len(shouldDeleteOlderProperties) > 0 {
+		if err := ps.remove(shouldDeleteOlderProperties); err != nil {
```

Comment:
To enhance performance, implement a repair operation in the data nodes. This operation receives the latest data, deletes all expired data, and then applies the latest one.